### DISCUSSION TITLE
NullReferenceException from MVC AppliationContextManager

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#3963] [Questions] [answered]
I'm using Csla 7.0.4 on .Net Framework 4.7.2, with ASP.Net MVC 5 / WebApi.  An API controller is calling some async data portal methods, but they are blowing up with a NullReferenceException from the Csla MVC ApplicationContextManager.  
```csharp
 at Csla.Web.Mvc.ApplicationContextManager.GetLocalContext()
 at Csla.ApplicationContext.get_LocalContext()
 at Csla.ApplicationContext.SetLogicalExecutionLocation(LogicalExecutionLocations location)
 at Csla.Server.DataPortal.ClearContext(DataPortalContext context)
 at Csla.Server.DataPortal.<Fetch>d__42.MoveNext()
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Csla.Channels.Local.LocalProxy.<Fetch>d__18.MoveNext()
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at Csla.Channels.Local.LocalProxy.<Fetch>d__18.MoveNext()
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Csla.DataPortal`1.<DoFetchAsync>d__19.MoveNext()
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Csla.DataPortal`1.<FetchAsync>d__24.MoveNext()
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
```

My configuration for Csla in the WebApi project
```chsarp
services
  .AddCsla(
    x => x.RegisterContextManager<Csla.Web.Mvc.ApplicationContextManager>()
	  .Data(
	    y => y.DefaultTransactionIsolationLevel(isolationLevel)
		      .DefaultTransactionTimeoutInSeconds(transactionTimeout)
			  .DefaultTransactionAsyncFlowOption(TransactionScopeAsyncFlowOption.Enabled)
	  )
	  .DataPortal(
	    y => y.AddServerSideDataPortal(
		  z => z.RegisterExceptionInspector<ExceptionInspector>()
		)
	)
);
```
The controller does this:

```
var reporterTask = GetReporterAsync(criteria);
var getTimeTask = serverTime.GetCurrentTimeInTimeZoneAsync("Eastern Standard Time");

await Task.WhenAll(reporterTask, getTimeTask);
```

We also away the actual call to the IDataPortal<T> in both cases (both of the calls above call our factory classes, which have the IDataPortal<T> injected.

Any ideas?  Should I be forcing Csla to use whatever the normal ApplicationContextManager is?

**Comment by rockfordlhotka** [comment]
Are your factory classes injected into the controller then? So they are created via DI, allowing the DI container to resolve `IDataPortal<T>`?

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Yes, everything is injected into the controller.  The controller & factories that take IDataPortal<T> are all transients as well.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka This seems like it could be related, but also makes things more confusing in some ways.  So until now I've been talking about our WebApi project.  We have another project, also WebApi we call remoting which servers as the http server for our Windows client application.  It only has two controllers, one of them being the Csla HttpPortalController.

We have integration tests, which run against IIS/IIS Express (locally) so that we have the BOs actually going through the real remote server and back.  Working on the Csla 7 update, I had just finished getting those tests all running and passing again.

A coworker though pulled down the same branch and tried running them, and some of the tests fail.  Not consistently, some alternate between passing and failing others always fail.  But its also ultimately an issue with HttpContext.Current becoming null in the Csla.Web.Mvc.ApplicationContextManager.  In her case though, this is happening trying to deserialize the UpdateRequest ReadOnlyBase that the Csla client builds and sends to the server.  So its not even getting tot he point of deserializing our code or really running anything of ours (e.g., no authorizer or exception inspector).   Here's the stack trace for that:

```
Csla.DataPortalException : {"Message":"An error has occurred.","ExceptionMessage":"Operation is not valid due to the current state of the object.","ExceptionType":"System.InvalidOperationException","StackTrace":"   at Remoting.RemotingContextManager.GetCurrentContext() in C:\\Source\\20\\branches\\CLEAN Csla\\Remoting\\RemotingContextManager.cs:line 97\r\n   at Remoting.RemotingContextManager.GetUser() in C:\\Source\\20\\branches\\CLEAN Csla\\Remoting\\RemotingContextManager.cs:line 33\r\n   at Csla.ReadOnlyBase`1.VerifyAuthorizationCache()\r\n   at Csla.ReadOnlyBase`1.CanReadProperty(IPropertyInfo property)\r\n   at Csla.ReadOnlyBase`1.CanReadProperty(IPropertyInfo property, Boolean throwOnFalse)\r\n   at Csla.ReadOnlyBase`1.GetProperty[P](PropertyInfo`1 propertyInfo, NoAccessBehavior noAccess)\r\n   at Csla.Server.Hosts.HttpPortalController.<InvokePortal>d__10.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Csla.Server.Hosts.HttpPortalController.<PostAsync>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Threading.Tasks.TaskHelpersExtensions.<CastToObject>d__1`1.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__1.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__15.MoveNext()"}
  ----> System.Net.WebException : The remote server returned an error: (500) Internal Server Error.
   at System.Net.WebClient.UploadDataInternal(Uri address, String method, Byte[] data, WebRequest& request)
   at System.Net.WebClient.UploadData(Uri address, String method, Byte[] data)
   at Csla.Channels.Http.HttpProxy.CallViaWebClient(Byte[] serialized, String operation, String routingToken)
   at Csla.BusinessBase`1.Save()
   at Test.BusinessTests.UserTest.UserPrivilegeTest() in C:\Source\20\branches\CLEAN Csla\Test\BusinessTests\UserTest.cs:line 72
--WebException
   at System.Net.WebClient.UploadDataInternal(Uri address, String method, Byte[] data, WebRequest& request)
   at System.Net.WebClient.UploadData(Uri address, String method, Byte[] data)
   at Csla.Channels.Http.HttpProxy.CallViaWebClient(Byte[] serialized, String operation, String routingToken)
```

**Reply by ajohnstone-ks** [reply]
Oh, the InvalidOperationException is from the MVC ApplicationContextManager; I just took your code for that, copied it into our remoting project and changed HttpContext.Current references to a method that check .Current for null and throw InvalidOperationException... just so we could set a breakpoint and see better in Csla where things were going wrong.

**Reply by rockfordlhotka** [reply]
So your code is running in an environment where HttpContext isn't available - at least that's what I'm hearing. That means you need to use a different application context manager, probably `Csla.Core.ApplicationContextManagerAsyncLocal'.

**Reply by rockfordlhotka** [reply]
If you are using a remote data portal, remember that the data portal works hard to provide you with a unified stack trace in the exception that includes server-side and client-side info all in one trace.

In other words, make sure you are troubleshooting the right thing: client or server. Maybe the exception isn't occurring on the web server in the controller (where HttpContext _should_ be valid), and might be occurring on the server where maybe you are hosting the code in a way where HttpContext isn't valid.

**Reply by ajohnstone-ks** [reply]
> So your code is running in an environment where HttpContext isn't available - at least that's what I'm hearing. That means you need to use a different application context manager, probably `Csla.Core.ApplicationContextManagerAsyncLocal'.

> If you are using a remote data portal, remember that the data portal works hard to provide you with a unified stack trace in the exception that includes server-side and client-side info all in one trace.
> 
> In other words, make sure you are troubleshooting the right thing: client or server. Maybe the exception isn't occurring on the web server in the controller (where HttpContext should be valid), and might be occurring on the server where maybe you are hosting the code in a way where HttpContext isn't valid.

We are both running under IIS Express, .Net framework 4.7.2.  On the client side (our NUnit tests) we are using `Csla.Core.ApplicationContextManagerStatic`.  In the remoting & WebApi projects (both MVC 5, .Net framework 4.7.2 running under IIS Express) we are using `Csla.Web.Mvc.ApplicationContextManager`  Those should be appropriate for each, correct?

The exception is definitely in the server side code; I've confirmed by checking the process id during debugging, and also looking at the ApplicationContext's ApplicationContextManager, and its using the Csla MVC version.  I assume nothing should change the context manager after initial configuration correct?

**Reply by ajohnstone-ks** [reply]
Something I discovered, but I'm not sure if it helps or how to get around it, but this happens the first async call we do in the controller.  I changed the call to be sync and things run up until the first async call.  Whatever is happening is before the DP code in my actual BO is entered.  Is there some Csla configuration I'm missing maybe?

**Reply by rockfordlhotka** [reply]
The first async call in the _client_ controller? Or in the server controller?

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka On the server; there's no controller on the client, its basically the same as a console app.

**Reply by rockfordlhotka** [reply]
oh, got it.

Is your server using a local or remote data portal?

**Reply by rockfordlhotka** [reply]
I was just going to try and replicate what you are doing, but I can't remember how to configure and use MVC5 - it has been too long.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka The server is in local data portal mode, it just talks directly to the DB.  Yes my next step was to try and create a repo outside of our solution.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka I was able to create a simple(ish) repo of the issue independently.  You can find it here:  https://github.com/ajohnstone-ks/WebApi

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Sorry, you can use Fidder to do this post

```
User-Agent: Fiddler
Authorization: APIKEY key="134"
Host: localhost:51423
Content-Length: 24
Content-Type: application/json

[{"CustomerId":"1234" }]
```

**Reply by ajohnstone-ks** [reply]
I'm playing with the sample repo myself to see if I can figure out what is tripping things up; so far I've removed Unity from the mix and just used the ServiceCollection/ServiceProvider, and that also has the same issue.  So I've rule out something with Unity.

**Reply by ajohnstone-ks** [reply]
So I've played with the sample I made, and it seems like with Csla 7 & MVC 5 on .Net framework doing any Async data portal operation will trigger the error.

**Reply by rockfordlhotka** [reply]
Repeatable errors are the easiest to debug

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka I've added the relevant Csla projects to my sample solution so I could debug things better, and it seems if I add this to my Csla configuration:
`.DataPortal(y => y.UseLocalProxy(z => z.FlowSynchronizationContext = true))`

The the sample run correctly.  I'm going to try it in my actual code, but does this seem like it should be the right solution to you?  Are there any pitfalls you're aware of in doing this?

**Reply by rockfordlhotka** [reply, marked as answer]
You know, I seem to remember that ASP.NET did use synchronization contexts, so that config change probably does make sense.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka So that setting, along with correctly configuring DependencyResolver & using it in my filters seems to have done the trick and my webapi is working now.  Thanks!
